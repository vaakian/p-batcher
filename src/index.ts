type Awaitable<T> = T | Promise<T>;
type BatchFn<Key, Result> = (keys: Key[]) => Awaitable<Result[]>;

interface BatchOptions {
  /**
   * The maximum number of keys to batch together in a single request.
   *
   * @default Infinity
   */
  maxBatchSize?: number;
  /**
   * A function that will be called to dispatch the batched request.
   *
   * @param fn
   * @returns
   */
  scheduler?: (fn: () => void) => void;
}

interface Task<Key, Result> {
  key: Key;
  resolve: (value: Result) => void;
  reject: (reason?: unknown) => void;
}

export function createPBatch<T, K>(
  batchFn: BatchFn<T, K>,
  options: BatchOptions
) {
  const {
    maxBatchSize: initialMaxBatchSize = Infinity,
    scheduler = (fn) => Promise.resolve().then(fn),
  } = options;

  let maxBatchSize = initialMaxBatchSize;
  const tasks: Task<T, K>[] = [];

  /**
   * Dispatches the current batch of keys.
   *
   * @returns
   */
  function dispatch() {
    if (tasks.length === 0) {
      return Promise.resolve([]);
    }

    const tasksToConsume = tasks.splice(0, maxBatchSize);
    const keys = tasksToConsume.map((t) => t.key);

    // TODO: Pass custom handler for errors
    return Promise.resolve(batchFn(keys))
      .then((results) =>
        tasksToConsume.forEach((task, i) => task.resolve(results[i]))
      )
      .catch((err) => tasksToConsume.forEach((task, i) => task.reject(err)));
  }

  /**
   * Gets a value from the cache, or fetches it from the source if it's not cached.
   *
   * @param key
   * @returns
   */
  function get(key: T) {
    return new Promise<K>((resolve, reject) => {
      tasks.push({ key, resolve, reject });
      if (tasks.length >= maxBatchSize) {
        return dispatch();
      }

      if (tasks.length === 1) {
        scheduler(() => dispatch());
      }
    });
  }

  get.setMaxBatchSize = (size: number) => {
    maxBatchSize = size;
  };

  return get;
}

export default createPBatch;