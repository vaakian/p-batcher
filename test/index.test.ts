import { test, expect, describe, vi } from "vitest"
import { createPBatch } from "../src"

describe("batcher", () => {
  const mockFetch = vi.fn((keys: number[]) => keys.map((k) => `response-${k}`))
  const get = createPBatch(mockFetch, { maxBatchSize: 3 })

  test("single call", async () => {
    expect(await get(1)).toBe("response-1")
  })

  test("multiple calls", async () => {
    const data = await Promise.all([1, 2, 3, 4, 5].map((k) => get(k)))

    expect(data).toEqual([
      "response-1",
      "response-2",
      "response-3",
      "response-4",
      "response-5",
    ])
    expect(mockFetch).toMatchSnapshot()
  })
})
