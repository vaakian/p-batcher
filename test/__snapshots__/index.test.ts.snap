// Vitest Snapshot v1

exports[`batcher > multiple calls 1`] = `
[MockFunction spy] {
  "calls": [
    [
      [
        1,
      ],
    ],
    [
      [
        1,
        2,
        3,
      ],
    ],
    [
      [
        4,
        5,
      ],
    ],
  ],
  "results": [
    {
      "type": "return",
      "value": [
        "response-1",
      ],
    },
    {
      "type": "return",
      "value": [
        "response-1",
        "response-2",
        "response-3",
      ],
    },
    {
      "type": "return",
      "value": [
        "response-4",
        "response-5",
      ],
    },
  ],
}
`;
